### DISCUSSION TITLE
How to approach (possibly) dangerous functionality?

### DISCUSSION THREAD

**Discussion by zc-devs** [woodpecker-ci/woodpecker#3758] [General]
#3655:
```go
EnvVars: []string{"WOODPECKER_BACKEND_K8S_ALLOW_NATIVE_SECRETS"}
```
#3609:
```go
EnvVars: []string{"WOODPECKER_BACKEND_K8S_POD_LABELS_ALLOW_FROM_STEP"},
EnvVars: []string{"WOODPECKER_BACKEND_K8S_POD_ANNOTATIONS_ALLOW_FROM_STEP"},
```

@qwerty287 wrote (https://github.com/woodpecker-ci/woodpecker/pull/3655#issuecomment-2146653633):
> I have some kind of issue with the kubernetes approach of options that could be/are dangerous.
>
> We currently have an env var option for everyone of them, just like WOODPECKER_BACKEND_K8S_ALLOW_NATIVE_SECRETS. The docker backend is doing it differently: You can't use the feature as long as the repo isn't trusted. If it's trusted you can use these features.
>
> I prefer the docker approach. It's configurable from ui and can be set per repo. Also, kubernetes and docker backends should work in a similar way and why do we have the trusted option if we don't use, but rather add new env options?
>
> Yes, I know that this is currently not really possible and we would have to do some refactoring for the backends first, but just wanted to point that out.

**Comment by zc-devs** [comment]
Treat this flags as [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/).
Yes, it can be used as security valve as "last" resort - disable this functionality on the instance, but this is not the main function.

In regard #3655. Security constraints are not implemented yet. But they [will be](https://github.com/woodpecker-ci/woodpecker/issues/3582#issuecomment-2096736572) by `Implement access controls` and `Implement secret masking` tasks. So, #3655 functionality is in experimental stage, I don't want to write documentation now, there is no need to announce it. Who want it, will enable the flag. When all tasks are implemented, we can change default value to `true`. Though, there will be "emergency brake" in case of some regression in this functionality. Note, that admin can decide whether it affects him or not.

In regard #3609. It is just precaution. Let admin decide 😆 

In regard #3538. This `user=0`, `group=0` are variation of `privileged=true`. `privileged` was implemented via `trusted` repo's flag. So, it does make sense to implement it like it was. So did I in #3538. Why didn't I make it before - cause I didn't pay much attention to this, I thought we could just patch it later if needed. Feature flag could save us, though ...


> kubernetes approach

Kubernetes features - Kubernetes flags, they are in `*BACKEND_K8S*` namespace ;) 

>We currently have an env var option... 
>`trusted` repo is configurable from ui and can be set per repo

While `trusted` repo can be set per repo, what if I want to set it per instance? What if I have 100500 repos in public instance?
I really like how Woodpecker configures via env vars. It is really "cloud native" and stateless. In contrast some apps use own config and just override them from envs. But they can support only subset of options via env vars, or set them only first time/run...

> You can't use the feature as long as the repo isn't trusted. If it's trusted you can use these features.
> why do we have the trusted option if we don't use, but rather add new env options?

It's just one flag. We need ~~go deeper~~ some permission/feature matrix then.
I need some K8s functionality (native secrets, for example). Someone want [volumes](https://github.com/woodpecker-ci/woodpecker/issues/3537#issuecomment-2016573502). We need more granular control, in case of emergency to not disable half of CI functionality. And to follow `least privileges` principle.

--- 

Overall, I like feature flags. Though, `trusted` should be improved.
Feature flags (instance wide) could coexist with feature/permission matrix (per repo). It would be the best, IMO.

**Reply by qwerty287** [reply]
> While trusted repo can be set per repo, what if I want to set it per instance?

It's not a big problem to add a var to enable it by default.

> What if I have 100500 repos in public instance?

I don't get that. Why do you want to make repos trusted on a public instance?

> Feature flags (instance wide) could coexist with feature/permission matrix (per repo). It would be the best, IMO.

Probably yes, as it looks like the kubernetes backend has a lot more features compared to docker.

But do you mean an "and" or "or" connection?

I.e. you're able to use a feature if it's enabled from env var *and* the repo is trusted? (That's what I would prefer)

I understand that you want to disable features globally, but still I think you shouldn't be able to use enabled features if the repo isn't trusted.
The kubernetes backend is, compared to docker, currently very unsuitable for public instances if you have a few repos that need additional features. They should work similar though.

Or did you mean by "matrix" that every feature can be enabled/disabled per repo?

**Comment by dominic-p** [comment]
Just to share my perspective here, my use case is a very private CI environment. Adding manual config in the UI to enable these features would add some friction for me.

Also, I agree with @zc-devs that the env variables can allow for a more fine grained approach than a blanket trusted switch. For example, an instances may need to allow setting security context but may not need secret access. In that case, we can have tighter controls using env vars.

> It's not a big problem to add a var to enable it by default.

This ideas concerns me a bit. If we adopt an "and" logic where in order to enable secret access (for example) you need both the env variable AND the repo to be trusted. I will probably wind up needing to make all of my repo's trusted. However, that will likely lead to a situation where I'm granting too many privileges and opening up additional security holes.

In the end, my feelings are that we need more granular control at the repo level (i.e. not just a single trusted switch) if we go with an "and" relationship between env vars and repo level permissions.

**Comment by zc-devs** [comment]
> Did you mean by "matrix" that every feature can be enabled/disabled per repo?

Yes. "My" feature gates are per instance. Matrix is per repo.
Also, feature flags do gate some backend functionality. Therefore it makes sense to do it in Agent as it is now. While "matrix" would be Server-side logic (UI, DB).

> It's not a big problem to add a var to enable it by default.

If you were talking about `flag per feature` it would be current feature flags :) But you are about `trusted repo`... It is not useful in current state, IMO ⬇️ 

> native secrets, volumes, etc...

Apart from me, it is explained by @dominic-p in https://github.com/woodpecker-ci/woodpecker/discussions/3758#discussioncomment-9668610 and @everflux in https://github.com/woodpecker-ci/woodpecker/issues/3537#issuecomment-2016573502.

> Why do you want to make repos trusted on a public instance?

I think this is a problem by itself. Why do I have to gain "root" access on the repos, if I want just enable some subset of features?

> But do you mean an "and" or "or" connection?
> I.e. you're able to use a feature if it's enabled from env var and the repo is trusted?

As I sad I do not consider `trusted` as a usable option at all...
Server's repo permission should be allowed to set by admin. Agent feature gate... is set by admin. As feature is not some thing to take precedence => I think logic is "or". You enable it per repo in Server UI `or` per instance via Agent's envs.

Edit: just to be more clear. In my thoughts, `trusted` transforms to permissions/features "matrix", instance-wide feature flags stays as envs in Agent.

**Reply by qwerty287** [reply]
I think I understood your points now, sorry if I was confusing.

> If you were talking about flag per feature it would be current feature flags :) But you are about trusted repo... It is not useful in current state, IMO ⬇️

I was talking about trusted.

> I think this is a problem by itself. Why do I have to gain "root" access on the repos, if I want just enable some subset of features?

I still don't understand your point here. You must have "root" access (I assume you mean woodpecker "admin" access) to enable these features and this is important if the feature could be dangerous. Any feature that isn't dangerous should be usable by anyone.

> As I sad I do not consider trusted as a usable option at all...
> Server's repo permission should be allowed to set by admin. Agent feature gate... is set by admin. As feature is not some thing to take precedence => I think logic is "or". You enable it per repo in Server UI or per instance via Agent's envs.

Yes, thanks, got it now.
However, it's not that easy to implement something like this because every backend has different features that could be dangerous. Implementing this is not that easy, the UI must differentiate between the backends.

The docker backend currently has these features only usable with trusted:
- privileged
- shm size
- dns
- dns search
- devices
- extra hosts
- network mode
- networks
- volumes
- tmpfs

(most of them are actually undocumented)

We could merge them to categories like "network" (dns, dns search, extra hosts, network mode, networks), "volumes" (devices, volumes, tmpfs), "resources" (shm size) and "security" (privileged).

I wouldn't allow that fine-granular control over the features, but rather allowing to enable/disable these categories to keep it simple and as backend-independent as possible.
What would you say about this? And which categories would be missing for a kubernetes environment (probably "secrets"?)?

And tbh I also don't really like the "or" connection. I'd rather add an env var that enables the feature by default (when a new repo is created). This way you can still opt-out some repos from the functionality.

And some off-topic question: Which of these undocumented features does kubernetes support? Because if kubernetes does not support them, we should probably move them to docker backend options. Privileged, extra hosts, volumes and networks are supported already. There are some more step features not requiring the repo to be trusted: MemSwapLimit, MemLimit, CPUQuota, CPUShares and CPUSet. Which of them can be supported by the kubernetes backend?

**Reply by zc-devs** [reply]
> You must have "root" access (I assume you mean woodpecker "admin" access)

I was talking about `privileged=true`="root" access (in container platform). It was about granularity.

> "admin" access to enable these features and this is important if the feature could be dangerous

I agreed on that ⬇️ 

> Server's repo permission should be allowed to set by admin

> While trusted repo can be set per repo, what if I want to set it per instance? What if I have 100500 repos in public instance?
> I still don't understand your point here

Seems, it was bad example... I'll try by example:
There is some instance with large number of repositories (1000). I (admin) have set 50 of them as trusted (or enable some feature). For some reason I need to revoke that permission.
1. Currently, there is no `trusted` label or filters in repo list. I need to go each repo settings and find out if it is `trusted` (1000 times).
2. I thought, I could disable Agent's feature flag on whole instance. Yes I still can and it partially works (see 3).
3. I see now, as there is `or` logic between Server-side permissions and Agent's flag, Agent's flag disables functionality only for repos, which uses permission/feature, but was not explicitly set `trusted`/feature-permission-enabled (on Server). For the last I still need do 1.

I suppose, I can use API, but... you know... 🤣 

> the UI must differentiate between the backends

Yes, there is Local backend now, which doesn't support the most of them, I guess.

> What would you say about this?

I like it. It is definitely better, than single `trusted`. Categories is good starting point, but there should be a way to extend it, I think.
I do not know about dns, dns search, but from my perspective: extra hosts is not so dangerous (if dangerous at all) as network mode.
BTW, what is "dangerous"? Dangerous to what, Docker host / Kubernetes Node?

> And tbh I also don't really like the "or" connection. I'd rather add an env var that enables the feature by default (when a new repo is created). This way you can still opt-out some repos from the functionality.

If permissions (categories) are implemented, then sure, it is an option. I'm fine with that.

But currently, I use Agent's flags also as experimental features switch. I think we should keep it somehow. Perhaps, we can use those `*INTERNAL*` vars and when feature stabilizes, just enable by default and drop the flag?

Another example of feature flags is [Keycloak](https://www.keycloak.org/server/features).

> Which of them can be supported by the kubernetes backend?

I'll be back later.

**Reply by qwerty287** [reply]
> Currently, there is no trusted label or filters in repo list. I need to go each repo settings and find out if it is trusted (1000 times).

Implementing something like this isn't a big problem.

> Categories is good starting point, but there should be a way to extend it, I think.

Shouldn't be hard to make this extensible I think.

> BTW, what is "dangerous"? Dangerous to what, Docker host / Kubernetes Node?

Yes, dangerous to the agent host.

> But currently, I use Agent's flags also as experimental features switch. I think we should keep it somehow. Perhaps, we can use those *INTERNAL* vars and when feature stabilizes, just enable by default and drop the flag?

Sure, we can use the env vars for now. Implementing this will take its time (from my point it would also be fine to just use regular names without `_INTERNAL_`).

**Reply by zc-devs** [reply]
| Feature                   | Supported | Implemented | Notes                                                                                                                                                                                                                                                                       | Links                                                                                                                                                                                                                             |
|---------------------------|-----------|-------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| privileged                | +         | +           |                                                                                                                                                                                                                                                                             |                                                                                                                                                                                                                                   |
| shm size                  | +         | -           | Shared memory can be set via in-memory-`emptyDir` volume as well as size.                                                                                                                                                                                                   |                                                                                                                                                                                                                                   |
| dns                       | +         | -           | Can be set via DNS Config                                                                                                                                                                                                                                                   | https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config                                                                                                                                           |
| dns search                | +         | -           | Can be set via DNS Config                                                                                                                                                                                                                                                   | https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config                                                                                                                                           |
| devices                   | +-        | +-          | Can be set as hostPath with privileged, or plugin + annotation. The first is unimplemented as we cannot mount volumes. The last, seems, works.                                                                                                                              | https://stackoverflow.com/questions/59290752/how-to-use-device-dev-video0-with-kubernetes,  https://github.com/kubernetes/kubernetes/issues/7890#issuecomment-766088805,  https://github.com/woodpecker-ci/woodpecker/issues/3188 |
| extra hosts               | +         | +           |                                                                                                                                                                                                                                                                             |                                                                                                                                                                                                                                   |
| network mode              | ?         | ?           | What is this? [Drivers](https://docs.docker.com/network/#drivers)? Then it controlled by [CNI implementation](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model). There is Node and Host ports, however. | https://stackoverflow.com/questions/63691946/kubernetes-what-is-hostport-and-hostip-used-for                                                                                                                                      |
| networks                  | +-        | -           | There is no direct alternative, as I know. Some CNI support Network policies.                                                                                                                                                                                               | https://kubernetes.io/docs/concepts/services-networking/network-policies/                                                                                                                                                         |
| volumes                   | +         | -           |                                                                                                                                                                                                                                                                             | https://github.com/woodpecker-ci/woodpecker/issues/1595                                                                                                                                                                           |
| tmpfs                     | +         | -           | Can be set via in-memory-`emptyDir` volume as well as size.                                                                                                                                                                                                                 |                                                                                                                                                                                                                                   |
|                           |           |             |                                                                                                                                                                                                                                                                             |                                                                                                                                                                                                                                   |
| MemLimit                  | +         | +           |                                                                                                                                                                                                                                                                             | https://woodpecker-ci.org/docs/administration/backends/kubernetes#resources                                                                                                                                                       |
| MemSwapLimit              | ?         | -           | Now Pod can use Swap in some cases, but it is not configurable per-Pod yet.                                                                                                                                                                                                 | https://stackoverflow.com/questions/63154923/how-to-add-swap-memory-in-kubernetes-pods,  https://kubernetes.io/blog/2023/08/24/swap-linux-beta/                                                                                   |
| CPUQuota CPUShares CPUSet | ?         | ?           | Isn't it covered by Reources?                                                                                                                                                                                                                                               | https://woodpecker-ci.org/docs/administration/backends/kubernetes#resources                                                                                                                                                       |

**Reply by zc-devs** [reply]
> Yes, dangerous to the agent host.

Then looks like dns, dns search and extra hosts are not dangerous. I may be wrong, but cannot imagine an example.

**Reply by qwerty287** [reply]
Thanks for this detailed list!

> network mode

I think it's https://docs.docker.com/network/#user-defined-networks. The different "modes" seem to be called "drivers" in the docker docs (also see https://stackoverflow.com/questions/74903232/what-is-the-difference-between-network-mode-and-networks-using-docker).

> MemLimit
> CPUQuota CPUShares CPUSet
> Isn't it covered by Reources?

Possible, yes. I just listed all advanced options that are not supported by the kubernetes backend.
However, these are directly defined on the step, while kubernetes resources are defined on the `backend_options`. The step-level MemLimit is also not supported by kubernetes.

If it would be possible to use both in kubernetes, that would be great. Otherwise they should be moved to `backend_options.docker` as well.

I.e. with docker you could do something like:
```yaml
steps:
  name: a
  image: b
  mem_limit: 1G
  cpu_shares: 5
```
This config wouldn't work with kubernetes right now.

> Then looks like dns, dns search and extra hosts are not dangerous. I may be wrong, but cannot imagine an example.

Yes, I was also wondering about this, I also don't think they're dangerous. I can check that later again.

**Reply by zc-devs** [reply]
> MemLimit

https://docs.docker.com/config/containers/resource_constraints/
https://serverfault.com/questions/753010/what-does-memory-reservation-flag-do-in-practice
https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits

Looks like K8s `requests.memory` = Docker `memory-reservation`, K8s `limits.memory` = Docker `memory`. So, there should not be a problem to implements it from step level.

> CPUQuota CPUShares CPUSet

https://github.com/moby/moby/issues/38839#issuecomment-470798734

Looks like K8s `limits.cpu` = Docker `cpus`. While for K8s guaranteed `requests.cpu` there is no direct alternative (?), but perhaps palying with `cpu-period` and `cpu-quota`.

> CPUSet

Docker's `cpuset-cpus` can be achieved by [Control CPU Management Policies on the Node](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy)


=> For CPU there is no direct equivalents
=> Probably, all resources should live in respective backends.

Anyway, Local doesn't support this (?). It is OK to have different syntax for different backends. Example - `image` in Local backend.

---

> Then looks like dns, dns search and extra hosts are not dangerous. I may be wrong, but cannot imagine an example.
> Yes, I was also wondering about this, I also don't think they're dangerous. I can check that later again.

Local backend have come up to my mind.
If it is implemented via `/etc/resolv.conf` and `/etc/hosts`, then it could just broke the host at least. In the worst case, it could trick host / other pipelines to connect to a malicious host.

**Reply by qwerty287** [reply]
> While for K8s guaranteed requests.cpu there is no direct alternative (?), but perhaps palying with cpu-period and cpu-quota.

We can also just move those that are not directly supported to backend options.

> Anyway, Local doesn't support this (?). It is OK to have different syntax for different backends. Example - image in Local backend.

Yes, but it's better to have as much as possible in the same way.

I know it's not ideal to have it in two different places, but it would be fine for me. Otherwise we probably could also just drop them as they are undocumented and I don't think anyone is using them

> > > Then looks like dns, dns search and extra hosts are not dangerous. I may be wrong, but cannot imagine an example.
> > Yes, I was also wondering about this, I also don't think they're dangerous. I can check that later again.
> 
> Local backend have come up to my mind.
> If it is implemented via /etc/resolv.conf and /etc/hosts, then it could just broke the host at least. In the worst case, it could trick > host / other pipelines to connect to a malicious host.

I don't think local should support this in general. You shouldn't run the agent using a privileged user so it shouldn't be possible to edit the files in `/etc`.

**Reply by zc-devs** [reply]
> I don't think local should support this in general. You shouldn't run the agent using a privileged user so it shouldn't be possible to edit the files in /etc.

Fair enough. Then `privileged` won't work either. The point is dangerousness of some features differs from backend to backend.

> I know it's not ideal to have it in two different places

If they are mutually exclusive like `requests.memory`, `limits.memory`, `limits.cpu` in step, but `requests.cpu` in backend options, I'm against this.
If duplicate them like `requests.memory`, `limits.memory`, `limits.cpu` in step and `requests.memory`, `limits.memory`, `limits.cpu`, `requests.cpu` in backend options, it is better, but worth to implement?

> it's better to have as much as possible in the same way

This universal options (like memory and CPU resources) would be different from their native options. If I want to use them I have to learn how they are translated into native options (even look into codebase maybe). It isn't convenient to person with Kubernetes background, nor Docker. Using backend options way more simple.
If I was using such options I would probably stick with backend options. But I do not use now, I use namespace's `LimitRange`. You mentioned, that Docker options are used not so often.

Therefore, question is: does it worth to implement universal resources options?
As CPUQuota CPUShares CPUSet are implemented already, I would move them into backend options and add documentations.

**Reply by qwerty287** [reply]
> This universal options (like memory and CPU resources) would be different from their native options. If I want to use them I have to learn how they are translated into native options (even look into codebase maybe). It isn't convenient to person with Kubernetes background, nor Docker. Using backend options way more simple.

Ah, I thought they would be usable in the same way for docker and kubernetes. In this case it's probably better to move it to backend options.

**Reply by zc-devs** [reply]
> they are undocumented

Apparently, [documented now](https://woodpecker-ci.org/docs/administration/server-config#woodpecker_limit_mem_swap) 😕

**Reply by qwerty287** [reply]
These are not the pipeline features (which you define in yaml). But yes, for the envs we probably need a deprecation first...

**Comment by qwerty287** [comment]
I started working on this in https://github.com/woodpecker-ci/woodpecker/pull/4025.

My roadmap:
1. Split trusted setting
2. Allow backends access to trusted configuration
3. Use the trusted setting in backend to check backend options
4. Check feature compatibility of native features and eventually move to docker backend options
5. Eventually deprecate old env vars to disable features

**Comment by qwerty287** [comment]
After my first PR to this we need to revisit it.

The main problem: You can now add repo/org agents and then, the admin agent should be able to control which repos get privileged access.

@anbraten suggests to fully remove the privileged part from the server and let the agent do everything. Problem: You maybe have multiple repos running on the same agent, some privileged, some not.

An option to merge these could be to set a list of repo IDs on the agent, so these repos are allowed to run privileged and the other ones not. What do you think about this?

If we decide here, I'm happy to continue my work on it.

**Reply by qwerty287** [reply]
@anbraten @zc-devs @woodpecker-ci/maintainers any opinions on this?

**Reply by zc-devs** [reply]
> @anbraten suggests to fully remove the privileged part from the server and let the agent do everything

https://github.com/woodpecker-ci/woodpecker/pull/4025#issuecomment-2382706698

---

So, the plan is
1. Get rid of Trusted options on Server,
2. Allow Agents/backends to implement security options (privileged, mounts etc.) in general and per repo (list),
3. Add option to Agent to restrict repositories, pipelines can run from?

I like this approach.

---

Also I would move `WOODPECKER_PRIVILEGED_PLUGINS` from Server to `WOODPECKER_PRIVILEGED_IMAGES` on Agent. And for this f...g BuildX plugin you would have to set `privileged:true` _explicitly_ in pipeline as well as `WOODPECKER_PRIVILEGED_IMAGES:docker.io/buildx` in Agent. Optionally restrict repositories.

---

#3539

1. Is there the docs?
2. Is that now usable only if I want dedicated Agent from resources point of view, and not usable from running "privileged" pipelines point of view?

**Reply by qwerty287** [reply]
> Is that now usable only if I want dedicated Agent from resources point of view, and not usable from running "privileged" pipelines point of view?

Currently, yes. You can't use a user-registered agent to run your pipeline with more privileges (that's basically why the discussion came up).
